includes/utils/class-amp-carousel.php

```diff
@@ -0,0 +1,160 @@
+	const FALLBACK_HEIGHT = 480;
+
+	/**
+	 * An object representation of the DOM.
+	 *
+	 * @var DOMDocument
+	 */
+	protected $dom;
+
+	/**
+	 * Instantiates the class.
+	 *
+	 * @param DOMDocument $dom The dom to use to create a carousel.
+	 */
+	public function __construct( $dom ) {
+		$this->dom = $dom;
+	}
+
+	/**
+	 * Creates and gets an <amp-carousel> with the given images and captions.
+	 *
+	 * @param array[] $images_and_captions {
+	 *     The images and their respective captions, if any.
+	 *
+	 *     @type DOMElement  $image A representation of an image, in index 0.
+	 *     @type string|null $caption The caption for that image, if any, in index 1.
+	 * }
+	 * @return DOMElement A representation of the <amp-carousel>.
+	 */
+	public function create_and_get( $images_and_captions ) {
+		$images = [];
+		foreach ( $images_and_captions as $image_and_caption ) {
+			$images[] = $image_and_caption[0];
+		}
+
+		list( $width, $height ) = $this->get_dimensions( $images );
+		$amp_carousel           = AMP_DOM_Utils::create_node(
+			$this->dom,
+			'amp-carousel',
+			[
+				'width'  => $width,
+				'height' => $height,
+				'type'   => 'slides',
+				'layout' => 'responsive',
+			]
+		);
+
+		foreach ( $images_and_captions as $image_and_caption ) {
+			$image   = $image_and_caption[0];
+			$caption = isset( $image_and_caption[1] ) ? $image_and_caption[1] : null;
+			$slide   = AMP_DOM_Utils::create_node(
+				$this->dom,
+				'div',
+				[ 'class' => 'slide' ]
+			);
+
+			// Ensure the image fills the entire <amp-carousel>, so the possible caption looks right.
+			if ( 'amp-img' === $image->tagName ) {
+				$image->setAttribute( 'layout', 'fill' );
+				$image->setAttribute( 'object-fit', 'cover' );
+			} elseif ( isset( $image->firstChild->tagName ) && 'amp-img' === $image->firstChild->tagName ) {
+				// If the <amp-img> is wrapped in an <a>.
+				$image->firstChild->setAttribute( 'layout', 'fill' );
+				$image->firstChild->setAttribute( 'object-fit', 'cover' );
+			}
+
+			$slide->appendChild( $image );
+
+			// If there's a caption, wrap it in a <div> and <span>, and append it to the slide.
+			if ( $caption ) {
+				$caption_wrapper = AMP_DOM_Utils::create_node(
+					$this->dom,
+					'div',
+					[ 'class' => 'amp-wp-gallery-caption' ]
+				);
+				$caption_span    = AMP_DOM_Utils::create_node( $this->dom, 'span', [] );
+				$text_node       = $this->dom->createTextNode( $caption );
+
+				$caption_span->appendChild( $text_node );
+				$caption_wrapper->appendChild( $caption_span );
+				$slide->appendChild( $caption_wrapper );
+			}
+
+			$amp_carousel->appendChild( $slide );
+		}
+
+		return $amp_carousel;
+	}
+
+	/**
+	 * Gets the carousel height by the containing images.
+	 *
+	 * @param array[] $images The images and captions to get the dimensions from.
+	 * @return array {
+	 *     Dimensions.
+	 *
+	 *     @type int $width  Width.
+	 *     @type int $height Height.
+	 * }
+	 */
+	public function get_dimensions( $images ) {
```

Comment:
This function is mainly moved from `AMP_Gallery_Block_Sanitizer::get_carousel_dimensions()`, with few changes.